wsgi/prz-recipe-options.rst

```diff
@@ -23,30 +18,131 @@ The default is to listen on 0.0.0.0:8080.
     listen directive in wsgi.ini (`listen = 0.0.0.0:`)
     Waitress will open on a random non-privileged port in this case.
 
-threads (vs zserver-threads)
-----------------------------
+threads
+-------
 
 The new ``threads`` option specifies the number of worker threads for both WSGI and ZServer.
 The WSGI default is 4 threads since this is the waitress default.
 The ZServer default of two threads remains unchanged.
+``zserver-threads`` can still be used to specify the number of worker threads for ZServer but not waitress.
+``zserver-threads`` is deprecated.
+
+.. note::
+
+    To assess the advantages and disadvantages of a particular server configuration, we need to understand some basics of how client requests are processed in a WSGI server.
+
+    Incoming client HTTP requests are often handled in an asynchronous way using the `asyncore <https://docs.python.org/3/library/asyncore.html#module-asyncore>`_ module from the Python standard library.
+    It was originally written by Sam Rushing as part of `Medusa <http://www.nightmare.com/medusa>`_ and is thus part of Zope's `ZServer <https://github.com/zopefoundation/ZServer/tree/master/src/ZServer>`_.
+    Its function in ``ZServer`` is to handle incoming client requests in an efficient way using the ``select`` and ``poll`` system calls.
+    I.e. it is responsible for clearing the request queue by dispatching incoming requests to available workers.
+    Only a **single** thread is needed for this task.
+    ``waitress`` uses ``asyncore`` just like ``ZServer`` for exactly the same task.
+    However ``asyncore`` is deprecated since Python 3.6 in favour of the new `asyncio <https://docs.python.org/3/library/asyncio.html#module-asyncio>`_ library using ``async/await`` syntax.
+    This is why ``waitress`` now ships with its own ``asyncore`` module named ``wasyncore``.
+    Both ``ZServer`` and ``waitress`` use "traditional" multi-threading based on the Python standard libraries' ``thread`` (renamed to ``_thread`` in Python 3) or ``threading`` modules to spawn worker threads.
+    The number of workers is configured in ``wsgi.ini`` (``waitress``) or ``zope.conf`` (``ZServer``).
+    Worker threads are spawned **once** when the Plone instance is starting up.
+    In Zope and Plone each of these workers will also open a ZODB connection.
+    You can check this in the ZMI Control Panel, "Database Management".
+
+    Other WSGI servers like ``bjoern``, ``gunicorn`` and ``uWSGI`` implement different or additional types of workers and they also differ in the way they are clearing the request queue.
+    We will try to cover at least some of the details in later chapters.
 
 wsgi
 ----
 
+This parameters has two forms.
+It can be either ``on`` or ``off`` signaling whether to use the default WSGI server ``waitress`` or not use a WSGI server but ``ZServer`` instead.
```

Comment:
Insert commas and a verb:

```
It can be either ``on`` or ``off``, signaling whether to use the default WSGI server ``waitress``, or not use a WSGI server but use ``ZServer`` instead.
```